src/compiler/checker.ts

```diff
@@ -15958,24 +15958,57 @@ namespace ts {
 
         function checkUnusedLocalsAndParameters(node: Node): void {
             if (node.parent.kind !== SyntaxKind.InterfaceDeclaration && noUnusedIdentifiers && !isInAmbientContext(node)) {
-                for (const key in node.locals) {
-                    const local = node.locals[key];
-                    if (!local.isReferenced) {
-                        if (local.valueDeclaration && getRootDeclaration(local.valueDeclaration).kind === SyntaxKind.Parameter) {
-                            const parameter = <ParameterDeclaration>getRootDeclaration(local.valueDeclaration);
-                            if (compilerOptions.noUnusedParameters &&
-                                !isParameterPropertyDeclaration(parameter) &&
-                                !parameterIsThisKeyword(parameter) &&
-                                !parameterNameStartsWithUnderscore(local.valueDeclaration.name)) {
-                                error(local.valueDeclaration.name, Diagnostics._0_is_declared_but_never_used, local.name);
-                            }
-                        }
-                        else if (compilerOptions.noUnusedLocals) {
-                            forEach(local.declarations, d => errorUnusedLocal(d.name || d, local.name));
+                if (compilerOptions.noUnusedParameters) {
+                    checkUnusedParameterLocals(node.locals);
+                }
+
+                if (compilerOptions.noUnusedLocals) {
+                    checkUnusedLocals(node.locals);
+                }
+            }
+        }
+
+        function checkUnusedParameterLocals(locals: Map<Symbol>): void {
+            let unusedParameterLocals: Symbol[] | undefined = [];

```

Comment:
From a performance perspective I was hesitant to make a new [] no matter what. Hence the slight added complexity.